### PR TITLE
Mean Normalisation Scaling

### PR DESCRIPTION
**Pull request by VascoSch92**
First version of the `MeanNormalizationScaling` as discussed in #763 

I create a new module scaling as discussed.

Probably, you will also add new scaling in this module.

### CODE REVIEW

File: feature_engine/scaling/mean_normalization.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# Authors: Vasco Schiavo <vasco.schiavo@protonmail.com>
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:1]: another protonmail fan, I like it!

File: feature_engine/scaling/mean_normalization.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+    ) -> None:
+
+        self.variables = _check_variables_input_value(variables)
+        # the following variable will be populated in the during the fit
+        self.params: Dict = {}
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:105]: in init, we can only have parameters that the user can tweak. So either we expose it, or we assign it in fit.

File: feature_engine/scaling/mean_normalization.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            variables to transform.
+
+        y: pandas Series, default=None
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+
+        for variable in self.variables_:
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:124]: i think the look is not optimal. X[self.variables_].mean() would return all means automatically, and the same for the below values. We should create a dictionary for mean, one for max and one for min, so then we can apply it at once to the entire dataframe without looping.

**Review by solegalli — COMMENTED** [review]
Hey @VascoSch92 

This is looking really good. Thank you for the first draft.

I think we could tidy it a bit so that we don't loop neither in fit nor in transform.

Could you take a look?

Thank you!

File: feature_engine/scaling/mean_normalization.py (hunk 2 above)
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:105]: I think it is nice if the user can see the fit parameters for the scaling, i.e., `mean`, `max`, `min`.

Or what do you think? Should I put it in the `fit` method?

Another question: private parameters ends with an underscore right? so it should be `params_` ?
I'm taking as inspiration `variables` and `variables_`.

File: feature_engine/scaling/mean_normalization.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        -------
+        X_new: pandas dataframe
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = self._check_transform_input_and_state(X)
+
+        # transformation
+        for variable in self.variables_:
```
[the remark this answers is not in the document]
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:152]: Yes this is true. The problem is that we have not just min max and mean but an also the dimension given by the columns. I think we can create a pandas data frame with all the info we need and then use it to do the transformations in a vectorised way.

File: feature_engine/scaling/mean_normalization.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            numerator = X[variable] - self.params[variable]["mean"]
+            denominator = self.params[variable]["max"] - self.params[variable]["min"]
+
+            # If max and min are equal, then the column's variable is constant.
+            # We set denominator to 1, and the column will be normalized to 0
+            if denominator == 0:
```
[the remark this answers is not in the document]
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:158]: As you want. I was following `Sklearn`. In fact, if you have zero denominator, then `Sklearn` don't thrown an error but divide by 1 (or don't divide).

I will change to raise an error ;-)

File: feature_engine/scaling/mean_normalization.py (hunk 2 above)
Comment (solegalli) [reply feature_engine/scaling/mean_normalization.py:105]: parameters ending in _ are learned during fit. That is sklearn convention. Private parameters start with _. So variables_ is learned during fit, and _variables would be private.

Yes, min,  max and min should be exposed to the user and learned in fit :)

**Comment by VascoSch92** [comment]
Hey @solegalli 

I addressed your comments.

Please let me know what you think :-)

File: feature_engine/scaling/mean_normalization.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            The training input samples. Can be the entire dataframe, not just the
+            variables to transform.
+
+        y: pandas Series, default=None
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+        self.params_ = pd.DataFrame(
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:121]: Do we need a dataframe? can we not make it work with dictionaries? Dataframes are heavier.

We'd need 3 dictionaries, one for mean, one for min and one for max. If we can make the code work with dictionaries, without transposing, this will be faster.

File: feature_engine/scaling/mean_normalization.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                "min": X[self.variables_].min(),
+            },
+        ).transpose()
+
+        # check for constant columns
+        constant_columns = self.params_.columns[
+            self.params_.loc["min"] == self.params_.loc["max"]
+        ].to_list()
+        if constant_columns:
+            raise ValueError(f"The following column/s are constant: {constant_columns}")
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:134]: can we make the error more informative? What's the problem with constant columns? We should try to squeeze that info in.

File: feature_engine/scaling/mean_normalization.py (hunk 6 above)
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:121]: Ah yes no problem. It should fast to change from `df` to `dict` ;)

File: feature_engine/scaling/mean_normalization.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+        y: pandas Series, default=None
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+        self.params_ = pd.DataFrame(
+            {
+                "mean": X[self.variables_].mean(),
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:123]: I am having second thoughts here. I think it is better to store the range (max-min), so that we make this computation once, and then we don't need to make it every time we transform the data. The transform will be faster. We should just store mean and range, if possible as dictionaries.

File: feature_engine/scaling/mean_normalization.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+@Substitution(
+    variables=_variables_numerical_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit=_fit_not_learn_docstring,
+    fit_transform=_fit_transform_docstring,
+    inverse_transform=_inverse_transform_docstring,
+)
+class MeanNormalizationScaling(BaseNumericalTransformer):
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:37]: let's use Scaler as name, MeanNormalizationScaler so we keep it like sklearn's names.

**Review by solegalli — COMMENTED** [review]
Hi @VascoSch92 

This is looking really good. The tests are great. 

Regarding the logic, I think we can speed this up if we store the range instead of min max,, and if we use dictionaries instead of dataframes. Could you give that a go?

Feel free to start working on the dosctrings and on adding a user guide :)

**Comment by VascoSch92** [comment]
Hey @solegalli 

 I changed what requested.
Now,

- `params_` is a dictionary `Dict[str, pd.Series]` with keys `'mean'` and '`range`'
- `MeanNormalizationScaling` -> `MeanNormalizationScaler`
- error message for constant columns is a little better :-)

File: feature_engine/scaling/mean_normalization.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        X_new: pandas dataframe
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = self._check_transform_input_and_state(X)
+
+        # transformation
+        X[self.variables_] = (
+            X[self.variables_] - self.params_["mean"][self.variables_]
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:157]: is the self.variables_  in `self.params_["mean"][self.variables_]`  necessary?

File: feature_engine/scaling/mean_normalization.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            The training input samples. Can be the entire dataframe, not just the
+            variables to transform.
+
+        y: pandas Series, default=None
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+        self.params_ = {
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:121]: this is OK. I think I still prefer having one dict for mean and one dict for range. Otherwise the dict might get too big and it might get confusing when the data has too many variables. Sklearn has 2 attributes, 1 for mean or center and one for scale or range, we would keep it consistent with their api as well if we create 2 attributes.

File: feature_engine/scaling/mean_normalization.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit=_fit_not_learn_docstring,
+    fit_transform=_fit_transform_docstring,
+    inverse_transform=_inverse_transform_docstring,
+)
+class MeanNormalizationScaler(BaseNumericalTransformer):
+    """
+    The MeanNormalizationScaler() applies the mean normalization scaling techinques
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:39]: Here I'd say that the meannormalizationscaler substracts the mean and then divides the result by the value range. And add a small formula of the trasformation

File: feature_engine/scaling/mean_normalization.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    to one or multuple columns of a dataframe.
+
+    Mean normalization is a way to implement feature scaling. Mean normalization
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:42]: Yes, more or less like this.

File: feature_engine/scaling/mean_normalization.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+    calculates and subtracts the mean of for every feature and divide this value
+    by the range (max - min).
+
+    A list of variables can be passed as an argument. Alternatively, the transformer
+    will automatically select and transform all variables of type numeric.
+
+    If a column is constant, then it will be transformed to the zero constant column.
+
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:50]: here we need to add a sentence that says more details in the user guide and create a user guide file with more detail about this transformation and an example with a real dataframe.

Here we create the user guide files: https://github.com/feature-engine/feature_engine/tree/main/docs/user_guide

We need a new folder for scaling, there an index, and then the file for the meannormalization scaler.

And also, to build the docs, we need to add a file to the api: https://github.com/feature-engine/feature_engine/tree/main/docs/api_doc  here we need a folder for scaling and in that folder an index, and the file for the scaler. 

If you check out the folder transformation inside those folders, you can very quickly get an idea of the files that we need and their format. The one for the api is super straightforward, the one for the user guide needs more work but I can support.

File: feature_engine/scaling/mean_normalization.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+    ) -> None:
+
+        self.variables = _check_variables_input_value(variables)
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        The method populate the variable `params_`.
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:107]: Could we replace with this:

Finds the mean and value range of each variable.

**Review by solegalli — COMMENTED** [review]
Hey @VascoSch92 really good work here. Thank you so much!

We need to add a few files to create the docs now. Would you be able to do that as well?

Thanks a lot for the hard work.

**Comment by solegalli** [comment]
Last but not least: we need to add the new module on the readme and on the frontpage of the documentation, which lives here: https://github.com/feature-engine/feature_engine/blob/main/docs/index.rst

Thank you!!

File: feature_engine/scaling/mean_normalization.py (hunk 10 above)
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:157]: I think yes. In this way you are taking on both side the columns in the right order

File: feature_engine/scaling/mean_normalization.py (hunk 11 above)
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py:121]: Ok so let's do `self.mean_` and `self.range_`

**Comment by VascoSch92** [comment]
Hey @solegalli 

- added the documentation for the new scaler
- divided `params_` into `mean_` and `range_`. Now `mean_` and `range_` are `pd.Series`. We can also make them `np.array` if you want. `Sklearn` uses np.arrays, should we make the same?

**Comment by VascoSch92** [comment]
Question: I added into the docstring this formula
```text
..math:: \frac{X_col - X_{mean}}{X_{max} - X_{min}}
```
is it supported by the docs of feature engine?

File: feature_engine/scaling/mean_normalization.py (hunk 10 above)
Comment (VascoSch92) [reply feature_engine/scaling/mean_normalization.py]: but I think that we can also do it with numpy array actually

File: docs/index.rst [hunk 16]
```diff
@@ -296,6 +296,15 @@ types and variable names match.
 - :doc:`api_doc/preprocessing/MatchCategories`: ensures categorical variables are of type 'category'
 - :doc:`api_doc/preprocessing/MatchVariables`: ensures that columns in test set match those in train set
 
+Scaling:
+~~~~~~~~
+
+Scaling the data can help to balance the impact of all variables on the model, and can improve the
```
Comment (solegalli) [inline docs/index.rst:302]:
```suggestion
Scaling the data can help to balance the impact of all variables on the model, and can improve 
```

File: feature_engine/scaling/mean_normalization.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+    Parameters
+    ----------
+    {variables}
+
+
+    Attributes
+    ----------
+    mean_:
+        a dictionary containing the mean, max and min of every given variable
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:59]:
```suggestion
        Dictionary containing the mean of the variables.
```

File: feature_engine/scaling/mean_normalization.py [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+
+    range_:
+        a pd.Series containing the range of the selected columns, i.e., the difference
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:62]:
```suggestion
        Dictionary containing the value range of of the variables.
```

File: feature_engine/scaling/mean_normalization.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        between the max and the min
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:63]:
```suggestion
```

File: feature_engine/scaling/mean_normalization.py [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+            The training input samples. Can be the entire dataframe, not just the
+            variables to transform.
+
+        y: pandas Series, default=None
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = super().fit(X)
+        self.mean_ = X[self.variables_].mean()
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:125]:
```suggestion
        self.mean_ = X[self.variables_].mean().to_dict()
```

File: feature_engine/scaling/mean_normalization.py [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+        self.range_ = X[self.variables_].max() - X[self.variables_].min()
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:126]: can we make this a dictionary? We use dictionaries in feature-engine :)

File: feature_engine/scaling/mean_normalization.py [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+        X_new: pandas dataframe
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = self._check_transform_input_and_state(X)
+
+        # transformation
+        X[self.variables_] = (
+            X[self.variables_] - self.mean_[self.variables_]
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:158]: If I remember from how I implemented pandas operations combined with dictionaries, pandas does automatically match keys with columns, so from memory, I don't think we need to pass self.variables_ to the dictionary. Could you test? remove it and see if the tests fail. Thank you!

File: feature_engine/scaling/mean_normalization.py [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+        ) / self.range_[self.variables_]
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:159]: same here

File: feature_engine/scaling/mean_normalization.py [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+        X_tr: pandas dataframe
+            The dataframe with the transformed variables.
+        """
+
+        # check input dataframe and if class was fitted
+        X = self._check_transform_input_and_state(X)
+
+        # inverse transform
+        X[self.variables_] = (
+            X[self.variables_] * self.range_[self.variables_]
```
Comment (solegalli) [inline feature_engine/scaling/mean_normalization.py:183]: same here

**Review by solegalli — COMMENTED** [review]
Hey @VascoSch92 thanks for the quick turnaround. The api docs look great.

**Comment by VascoSch92** [comment]
Hey @solegalli 

I changed to dictionaries instead of `pd.Series`, and it works :-)

**Comment by solegalli** [comment]
Amazing! Thanks a lot!

We just need to add a description / demo to the user guide folder in the docs and we are good to go then :)

**Comment by VascoSch92** [comment]
@solegalli quick question: in general feature engine use scaling transformers from sklearn. Should we also include examples of these transformers?

**Comment by VascoSch92** [comment]
Hey @solegalli 

I updated the docs with a demo.

Let me know what do you think. It is just a first version :-)

**Comment by VascoSch92** [comment]
Hey @solegalli :-) did you have time to look at the latest changes?

**Comment by solegalli** [comment]
Pending acceptance of suggested changes: https://github.com/VascoSch92/feature_engine/pull/1
